src/freenet/client/filter/WAVFilter.java

```diff
@@ -0,0 +1,131 @@
+			}
+			if(size != FMT_SIZE_BASIC && size != FMT_SIZE_cbSize && size != FMT_SIZE_cbSize_extension) {
+				throw new DataFilterException(l10n("invalidTitle"), l10n("invalidTitle"), "fmt chunk size is invalid");
+			}
+			ctx.format = Short.reverseBytes(input.readShort());
+			if(ctx.format != WAVE_FORMAT_PCM && ctx.format != WAVE_FORMAT_IEEE_FLOAT && ctx.format != WAVE_FORMAT_ALAW && ctx.format != WAVE_FORMAT_MULAW) {
+				throw new DataFilterException(l10n("invalidTitle"), l10n("invalidTitle"), "WAV file uses a not yet supported format");
+			}
+			ctx.nChannels = Short.reverseBytes(input.readShort());
+			output.write(ID);
+			writeLittleEndianInt(output, size);
+			output.writeInt((Short.reverseBytes((short) ctx.format) << 16) | Short.reverseBytes((short) ctx.nChannels));
+			ctx.nSamplesPerSec = readLittleEndianInt(input);
+			writeLittleEndianInt(output, ctx.nSamplesPerSec);
+			int nAvgBytesPerSec = readLittleEndianInt(input);
+			writeLittleEndianInt(output, nAvgBytesPerSec);
+			ctx.nBlockAlign = Short.reverseBytes(input.readShort());
+			ctx.wBitsPerSample = Short.reverseBytes(input.readShort());
+			output.writeInt((Short.reverseBytes((short) ctx.nBlockAlign) << 16) | Short.reverseBytes((short) ctx.wBitsPerSample));
+			ctx.hasfmt = true;
+			if(size > FMT_SIZE_BASIC) {
+				short cbSize = Short.reverseBytes(input.readShort());
+				if(cbSize + FMT_SIZE_cbSize != size) {
+					throw new DataFilterException(l10n("invalidTitle"), l10n("invalidTitle"), "fmt chunk size is invalid");
+				}
+				output.writeShort(Short.reverseBytes(cbSize));
+			}
+			if(size > FMT_SIZE_cbSize) {
+				// wValidBitsPerSample, dwChannelMask, and SubFormat GUID
+				passthroughBytes(input, output, FMT_SIZE_cbSize_extension - FMT_SIZE_cbSize);
+			}
+			// Further checks
+			if((ctx.format == WAVE_FORMAT_ALAW || ctx.format == WAVE_FORMAT_MULAW) && ctx.wBitsPerSample != 8) {
+				// These formats are 8-bit
+				throw new DataFilterException(l10n("invalidTitle"), l10n("invalidTitle"), "Unexpected bits per sample value");
+			}
+			return;
+		}
+		if(!ctx.hasfmt) {
+			throw new DataFilterException(l10n("invalidTitle"), l10n("invalidTitle"), "Unexpected header chunk was encountered, instead of fmt chunk");
+		}
+		if(ID[0] == 'd' && ID[1] == 'a' && ID[2] == 't' && ID[3] == 'a') {
+			// audio data
+			output.write(ID);
+			writeLittleEndianInt(output, size);
+			passthroughBytes(input, output, size);
+			if((size & 1) != 0) { // Add padding if necessary
+				output.writeByte(input.readByte());
+			}
+			ctx.hasdata = true;
+		} else if(ID[0] == 'f' && ID[1] == 'a' && ID[2] == 'c' && ID[3] == 't') {
+			if(size < 4) {
+				throw new DataFilterException(l10n("invalidTitle"), l10n("invalidTitle"), "fact chunk must contain at least 4 bytes");
+			}
+			// Just dwSampleLength (Number of samples) here, pass through
+			output.write(ID);
+			writeLittleEndianInt(output, size);
+			passthroughBytes(input, output, size);

```

Comment:
The passed-through bytes are format-specific. This would allow passing arbitrary unverified non-audio data.
However, the currently supported formats are uncompressed and should, if at all, only have the first `dwSampleLength` field.
Can we limit the number of passed-through bytes to 4 and ignore whichever bytes remain? Of course this should be adjusted accordingly in the size following the `fmt ` chunk ID.

Reply:
limiting the number of passed-through bytes doesn't guarantee the remaining 00 bytes won't have an effect, so I just discard this entire trunk when the size is not 4.